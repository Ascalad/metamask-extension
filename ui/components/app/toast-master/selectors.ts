import { isEvmAccountType } from '@metamask/keyring-api';
import { InternalAccount } from '@metamask/keyring-internal-api';
import { getAlertEnabledness } from '../../../ducks/metamask/metamask';
import {
  SURVEY_DATE,
  SURVEY_END_TIME,
  SURVEY_START_TIME,
} from '../../../helpers/constants/survey';
import { getPermittedAccountsForCurrentTab } from '../../../selectors';
import { MetaMaskReduxState } from '../../../store/store';
import { getIsPrivacyToastRecent } from './utils';

// TODO: get this into one of the larger definitions of state type
type State = Omit<MetaMaskReduxState, 'appState'> & {
  appState: {
    showNftDetectionEnablementToast?: boolean;
    ///: BEGIN:ONLY_INCLUDE_IF(multi-srp)
    showNewSrpAddedToast?: boolean;
    ///: END:ONLY_INCLUDE_IF
  };
  metamask: {
    newPrivacyPolicyToastClickedOrClosed?: boolean;
    newPrivacyPolicyToastShownDate?: number;
    onboardingDate?: number;
    showNftDetectionEnablementToast?: boolean;
    surveyLinkLastClickedOrClosed?: number;
    switchedNetworkNeverShowMessage?: boolean;
    // Keep the remoteFeatureFlags property as is
    remoteFeatureFlags?: Record<string, unknown>;
  };
};

/**
 * Determines if the survey toast should be shown based on the current time, survey start and end times, and whether the survey link was last clicked or closed.
 *
 * @param state - The application state containing the necessary survey data.
 * @returns True if the current time is between the survey start and end times and the survey link was not last clicked or closed. False otherwise.
 */
export function selectShowSurveyToast(state: State): boolean {
  if (state.metamask?.surveyLinkLastClickedOrClosed) {
    return false;
  }

  const startTime = new Date(`${SURVEY_DATE} ${SURVEY_START_TIME}`).getTime();
  const endTime = new Date(`${SURVEY_DATE} ${SURVEY_END_TIME}`).getTime();
  const now = Date.now();

  return now > startTime && now < endTime;
}

/**
 * Determines if the privacy policy toast should be shown based on the remote feature flag and whether the toast was clicked or closed.
 *
 * @param state - The application state containing the privacy policy data.
 * @returns Boolean is True if the toast should be shown, and the number is the date the toast was last shown.
 */
export function selectShowPrivacyPolicyToast(state: State): {
  showPrivacyPolicyToast: boolean;
  newPrivacyPolicyToastShownDate?: number;
} {
  const {
    newPrivacyPolicyToastClickedOrClosed,
    newPrivacyPolicyToastShownDate,
    onboardingDate,
  } = state.metamask || {};

  const remoteFeatureFlags = state.metamask?.remoteFeatureFlags || {};

  const policyUpdateDate = String(
    remoteFeatureFlags.transactionsPrivacyPolicyUpdate || '',
  );

  // If the feature flag isn't set or is empty, don't show the toast
  if (!policyUpdateDate) {
    // No policy update date found
    return { showPrivacyPolicyToast: false, newPrivacyPolicyToastShownDate };
  }

  // Create dates and ensure consistent timezone handling
  const newPrivacyPolicyDate = new Date(policyUpdateDate);
  const currentDate = new Date();

  const currentTimestamp = currentDate.getTime();
  const policyTimestamp = newPrivacyPolicyDate.getTime();

  const isRecent = getIsPrivacyToastRecent(newPrivacyPolicyToastShownDate);
  const beforePolicyUpdate =
    !onboardingDate || onboardingDate <= policyTimestamp;
  const afterPolicyDate = currentTimestamp >= policyTimestamp;

  const showPrivacyPolicyToast =
    !newPrivacyPolicyToastClickedOrClosed &&
    isRecent &&
    beforePolicyUpdate &&
    afterPolicyDate;

  return { showPrivacyPolicyToast, newPrivacyPolicyToastShownDate };
}

// If there is more than one connected account to activeTabOrigin,
// *BUT* the current account is not one of them, show the banner
export function selectShowConnectAccountToast(
  state: State,
  account: InternalAccount,
): boolean {
  const allowShowAccountSetting = getAlertEnabledness(state).unconnectedAccount;
  const connectedAccounts = getPermittedAccountsForCurrentTab(state);
  const isEvmAccount = isEvmAccountType(account?.type);

  return (
    allowShowAccountSetting &&
    account &&
    state.activeTab?.origin &&
    isEvmAccount &&
    connectedAccounts.length > 0 &&
    !connectedAccounts.some((address) => address === account.address)
  );
}

/**
 * Retrieves user preference to never see the "Switched Network" toast
 *
 * @param state - Redux state object.
 * @returns Boolean preference value
 */
export function selectSwitchedNetworkNeverShowMessage(state: State): boolean {
  return Boolean(state.metamask.switchedNetworkNeverShowMessage);
}

/**
 * Retrieves user preference to see the "New SRP Added" toast
 *
 * @param state - Redux state object.
 * @returns Boolean preference value
 */
///: BEGIN:ONLY_INCLUDE_IF(multi-srp)
export function selectNewSrpAdded(state: State): boolean {
  return Boolean(state.appState.showNewSrpAddedToast);
}
///: END:ONLY_INCLUDE_IF

export function selectNftDetectionEnablementToast(state: State): boolean {
  return Boolean(state.appState?.showNftDetectionEnablementToast);
}
