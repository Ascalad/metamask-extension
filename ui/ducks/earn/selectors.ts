import { EarnControllerState } from '@metamask/earn-controller';
import { createSelector } from 'reselect';

export type EarnAppState = {
  metamask: EarnControllerState;
};

const getPooledStakingState = (state: EarnAppState) => {
  console.log('state: ', state);
  return state.metamask.pooled_staking;
};

export const getPooledStakingEligibility = createSelector(
  getPooledStakingState,
  (pooledStakingState) => pooledStakingState.isEligible,
);
