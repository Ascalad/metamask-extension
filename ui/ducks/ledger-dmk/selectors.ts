import type { LedgerDmkState } from './ledger-dmk';

export const getLedgerConnectedDevice = (state: LedgerDmkState) =>
  state.connectedDevice;

export const getLedgerDeviceStatus = (state: LedgerDmkState) =>
  state.deviceStatus;

export const getLedgerEthSigner = (state: LedgerDmkState) => state.ethSigner;

export const getLedgerDmk = (state: LedgerDmkState) => state.dmk;

export const getLedgerSessionId = (state: LedgerDmkState) => state.sessionId;

export const getLedgerTransportType = (state: LedgerDmkState) =>
  state.transportType;
