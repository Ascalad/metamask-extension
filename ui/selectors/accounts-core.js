import { SubjectType } from '@metamask/permission-controller';
import { memoize } from 'lodash';

export function getPinnedAccountsList(state) {
  return state.metamask.pinnedAccountList;
}

export function getHiddenAccountsList(state) {
  return state.metamask.hiddenAccountList;
}

export function getInternalAccount(state, accountId) {
  return state.metamask.internalAccounts.accounts[accountId];
}

export function getOriginOfCurrentTab(state) {
  return state.activeTab.origin;
}
/**
 * Get account balances state.
 *
 * @param {object} state - Redux state
 * @returns {object} A map of account addresses to account objects (which includes the account balance)
 */
export function getMetaMaskAccountBalances(state) {
  return state.metamask.accounts;
}

export function getSubjectMetadata(state) {
  return state.metamask.subjectMetadata;
}

export function getMetaMaskKeyrings(state) {
  return state.metamask.keyrings;
}
/**
 * @param {string} svgString - The raw SVG string to make embeddable.
 * @returns {string} The embeddable SVG string.
 */
const getEmbeddableSvg = memoize(
  (svgString) => `data:image/svg+xml;utf8,${encodeURIComponent(svgString)}`,
);

export function getTargetSubjectMetadata(state, origin) {
  const metadata = getSubjectMetadata(state)[origin];

  if (metadata?.subjectType === SubjectType.Snap) {
    const { svgIcon, ...remainingMetadata } = metadata;
    return {
      ...remainingMetadata,
      iconUrl: svgIcon ? getEmbeddableSvg(svgIcon) : null,
    };
  }

  return metadata;
}
