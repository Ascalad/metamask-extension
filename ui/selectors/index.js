export * from '../pages/confirmations/selectors/confirm';
export * from './confirm-transaction';
export * from './custom-gas';
export * from './first-time-flow';
export * from './metametrics';
export * from './permissions';
export * from './selectors';
export * from './transactions';
export * from './approvals';
export * from './accounts';
export * from './getMetaMaskAccounts';
export * from './getMetaMaskAccountsConnected';
export * from './accounts-for-tab';
export * from './getMetaMaskAccountsOrdered';
export * from './getMultichainBalances';
export * from './getMultichainNetworkProviders';
export * from './getMultichainSelectedAccountCachedBalance';
export * from './getMultichainIsEvm';
export * from './getMultichainCurrentNetwork';
export * from './getMultichainProviderConfig';
export * from './getMultichainNetwork';
export * from './accounts-core';
