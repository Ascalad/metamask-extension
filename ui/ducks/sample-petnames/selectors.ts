import type { Hex } from '@metamask/utils';
import { createSelector } from 'reselect';
import { getCurrentChainId } from '../../../shared/modules/selectors/networks';
import type { MetaMaskReduxState } from '../../store/store';
// eslint-disable-next-line import/no-restricted-paths
import { type SamplePetnamesControllerState } from '../../../app/scripts/controllers/sample/sample-petnames-controller';

// Selectors
export const getSamplePetnamesByChainIdAndAddress = (
  state: Pick<MetaMaskReduxState, 'metamask'>,
) =>
  state.metamask.samplePetnamesByChainIdAndAddress as
    | SamplePetnamesControllerState['samplePetnamesByChainIdAndAddress']
    | undefined;

export const getPetnamesForCurrentChain = createSelector(
  [getSamplePetnamesByChainIdAndAddress, getCurrentChainId],
  (petnamesByChainIdAndAddress, chainId): Record<Hex, string> =>
    petnamesByChainIdAndAddress?.[chainId] ?? {},
);
