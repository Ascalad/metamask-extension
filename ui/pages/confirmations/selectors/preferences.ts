import { MetaMaskSliceControllerState } from '../../../ducks/metamask/metamask';
import { getPreferences } from '../../../selectors';

export const selectUseTransactionSimulations = (
  state: MetaMaskSliceControllerState<'PreferencesController'>,
) => state.metamask.PreferencesController.useTransactionSimulations;

export function selectConfirmationAdvancedDetailsOpen(
  state: Parameters<typeof getPreferences>[0],
) {
  return Boolean(getPreferences(state).showConfirmationAdvancedDetails);
}
