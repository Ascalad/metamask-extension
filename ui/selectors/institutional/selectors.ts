import { toChecksumAddress } from 'ethereumjs-util';
import { getAccountType } from '../selectors';
import { AccountsState, getSelectedInternalAccount } from '../accounts';
import {
  ProviderConfigState,
  getProviderConfig,
} from '../../../shared/modules/selectors/networks';
import { hexToDecimal } from '../../../shared/modules/conversion.utils';
// TODO: Remove restricted import
// eslint-disable-next-line import/no-restricted-paths
import { normalizeSafeAddress } from '../../../app/scripts/lib/multichain/address';
import { AccountType } from '../../../shared/constants/custody';
import { BackgroundStateProxy } from '../../../shared/types/metamask';
import { MetaMaskReduxState } from '../../store/store';

export function getWaitForConfirmDeepLinkDialog(state: {
  metamask: Pick<BackgroundStateProxy, 'CustodyController'>;
}) {
  return state.metamask.CustodyController.waitForConfirmDeepLinkDialog;
}

export function getTransactionStatusMap(state: {
  metamask: Pick<BackgroundStateProxy, 'CustodyController'>;
}) {
  return state.metamask.CustodyController.custodyStatusMaps;
}

export function getCustodyAccountDetails(state: {
  metamask: Pick<BackgroundStateProxy, 'CustodyController'>;
}) {
  return state.metamask.CustodyController.custodyAccountDetails;
}

export function getCustodyAccountSupportedChains(
  state: {
    metamask: Pick<BackgroundStateProxy, 'CustodyController'>;
  },
  address: string,
): { supportedChains: string[] } | undefined {
  const chains = state.metamask.CustodyController.custodianSupportedChains
    ? state.metamask.CustodyController.custodianSupportedChains[
        toChecksumAddress(address)
      ]
    : undefined;

  if (chains && 'supportedChains' in chains) {
    return chains;
  }

  return undefined;
}

export function getMmiPortfolioEnabled(state: {
  metamask: Pick<BackgroundStateProxy, 'MmiConfigurationController'>;
}) {
  if (process.env.IN_TEST) {
    return true;
  }

  return state.metamask.MmiConfigurationController.mmiConfiguration?.portfolio
    ?.enabled;
}

export function getMmiPortfolioUrl(state: {
  metamask: Pick<BackgroundStateProxy, 'MmiConfigurationController'>;
}) {
  return (
    state.metamask.MmiConfigurationController.mmiConfiguration?.portfolio
      ?.url || ''
  );
}

export function getConfiguredCustodians(state: {
  metamask: Pick<BackgroundStateProxy, 'MmiConfigurationController'>;
}) {
  return (
    state.metamask.MmiConfigurationController.mmiConfiguration?.custodians || []
  );
}

export function getCustodianIconForAddress(
  state: {
    metamask: Pick<
      BackgroundStateProxy,
      'CustodyController' | 'MmiConfigurationController'
    >;
  },
  address: string,
) {
  let custodianIcon;

  const checksummedAddress = address && normalizeSafeAddress(address);
  if (
    checksummedAddress &&
    state.metamask.CustodyController.custodyAccountDetails?.[checksummedAddress]
  ) {
    const { custodianName } =
      state.metamask.CustodyController.custodyAccountDetails[
        checksummedAddress
      ];
    custodianIcon =
      state.metamask.MmiConfigurationController.mmiConfiguration?.custodians?.find(
        (custodian) => custodian.envName === custodianName,
      )?.iconUrl;
  }

  return custodianIcon;
}

export function getIsCustodianSupportedChain(
  state: {
    metamask: Pick<BackgroundStateProxy, 'CustodyController'>;
  } & AccountsState &
    ProviderConfigState,
) {
  try {
    const selectedAccount = getSelectedInternalAccount(state);
    const accountType = getAccountType(state);

    if (!selectedAccount || !accountType) {
      throw new Error('Invalid state');
    }

    const providerConfig = getProviderConfig(state);

    if (typeof providerConfig.chainId !== 'string') {
      throw new Error('Chain ID must be a string');
    }

    // eslint-disable-next-line require-unicode-regexp
    if (!/^0x[0-9a-f]+$/i.test(providerConfig.chainId)) {
      throw new Error('Chain ID must be a hexadecimal number');
    }

    if (accountType !== AccountType.CUSTODY) {
      return true;
    }

    const supportedChains = getCustodyAccountSupportedChains(
      state,
      selectedAccount.address,
    );

    if (!supportedChains?.supportedChains) {
      return true;
    }

    return supportedChains.supportedChains.includes(
      hexToDecimal(providerConfig.chainId),
    );
  } catch (error) {
    console.error(error);
    throw error;
  }
}

export function getMMIAddressFromModalOrAddress(
  state: {
    metamask: Pick<BackgroundStateProxy, 'MmiConfigurationController'>;
  } & AccountsState &
    Pick<MetaMaskReduxState, 'appState'>,
) {
  const modalAddress = state?.appState?.modal?.modalState?.props?.address;
  const selectedAddress = getSelectedInternalAccount(state)?.address;

  return modalAddress || selectedAddress;
}

export function getMMIConfiguration(state: {
  metamask: Pick<BackgroundStateProxy, 'MmiConfigurationController'>;
}) {
  return state.metamask.MmiConfigurationController.mmiConfiguration ?? {};
}

export function getInteractiveReplacementToken(state: {
  metamask: Pick<BackgroundStateProxy, 'AppStateController'>;
}) {
  return state.metamask.AppStateController.interactiveReplacementToken ?? {};
}

export function getCustodianDeepLink(state: {
  metamask: Pick<BackgroundStateProxy, 'AppStateController'>;
}) {
  return state.metamask.AppStateController.custodianDeepLink ?? {};
}

export function getIsNoteToTraderSupported(
  state: {
    metamask: Pick<
      BackgroundStateProxy,
      'CustodyController' | 'MmiConfigurationController'
    >;
  },
  fromChecksumHexAddress: string,
) {
  const {
    CustodyController: { custodyAccountDetails },
    MmiConfigurationController: { mmiConfiguration },
  } = state.metamask;
  const accountDetails = custodyAccountDetails?.[fromChecksumHexAddress];

  if (!accountDetails) {
    return false;
  }

  const foundCustodian = mmiConfiguration?.custodians?.find(
    (custodian) => custodian.envName === accountDetails.custodianName,
  );

  return foundCustodian ? foundCustodian.isNoteToTraderSupported : false;
}

export function getIsCustodianPublishesTransactionSupported(
  state: {
    metamask: Pick<
      BackgroundStateProxy,
      'CustodyController' | 'MmiConfigurationController'
    >;
  },
  fromChecksumHexAddress: string,
) {
  const {
    CustodyController: { custodyAccountDetails },
    MmiConfigurationController: { mmiConfiguration },
  } = state.metamask;
  const accountDetails = custodyAccountDetails?.[fromChecksumHexAddress];

  if (!accountDetails) {
    return false;
  }

  const foundCustodian = mmiConfiguration?.custodians?.find(
    (custodian) => custodian.envName === accountDetails.custodianName,
  );

  return foundCustodian ? foundCustodian.custodianPublishesTransaction : false;
}

export function getNoteToTraderMessage(state: {
  metamask: Pick<BackgroundStateProxy, 'AppStateController'>;
}) {
  return state.metamask.AppStateController.noteToTraderMessage ?? '';
}
