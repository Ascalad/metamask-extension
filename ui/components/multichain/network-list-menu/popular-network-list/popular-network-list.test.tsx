import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { useDispatch, useSelector } from 'react-redux';
import configureStore from 'redux-mock-store';
import { renderWithProvider } from '../../../../../test/lib/render-helpers';
import { getUnapprovedConfirmations } from '../../../../selectors';
import PopularNetworkList from './popular-network-list';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

const STATE_MOCK = {
  metamask: {
    providerConfig: {
      chainId: '0x1',
    },
  },
};

describe('PopularNetworkList', () => {
  const store = configureStore()(STATE_MOCK);
  const useDispatchMock = useDispatch as jest.Mock;
  const useSelectorMock = useSelector as jest.Mock;
  const mockDispatch = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();
    useDispatchMock.mockReturnValue(mockDispatch);

    useSelectorMock.mockImplementation((selector) => {
      if (selector === getUnapprovedConfirmations) {
        return [];
      }
      return undefined;
    });
  });

  const defaultProps = {
    searchAddNetworkResults: [],
  };

  it('renders popular list component', () => {
    const { container } = renderWithProvider(
      <PopularNetworkList {...defaultProps} />,
      store,
    );

    expect(container).toMatchSnapshot();
  });

  it('displays the edge case message when there are no networks', () => {
    render(<PopularNetworkList {...defaultProps} />);
    expect(screen.getByTestId('all-networks-added')).toBeInTheDocument();
  });

  it('displays the network list when networks are provided', () => {
    const props = {
      ...defaultProps,
      searchAddNetworkResults: [
        {
          chainId: '1',
          nickname: 'Network 1',
          rpcPrefs: { imageUrl: 'https://example.com/image1.png' },
        },
        {
          chainId: '2',
          nickname: 'Network 2',
          rpcPrefs: { imageUrl: 'https://example.com/image2.png' },
        },
      ],
    };

    render(<PopularNetworkList {...props} />);
    expect(screen.getByText('Network 1')).toBeInTheDocument();
    expect(screen.getByText('Network 2')).toBeInTheDocument();
  });
});
