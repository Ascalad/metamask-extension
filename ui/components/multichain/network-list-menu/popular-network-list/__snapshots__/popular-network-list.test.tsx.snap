// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopularNetworkList renders popular list component 1`] = `
<div>
  <div
    class="mm-box add-network__edge-case-box mm-box--margin-top-4 mm-box--margin-right-6 mm-box--margin-left-6 mm-box--padding-4 mm-box--display-flex mm-box--flex-direction-row mm-box--background-color-background-alternative mm-box--rounded-md"
  >
    <div
      class="mm-box mm-box--margin-right-4"
    >
      <img
        src="images/info-fox.svg"
      />
    </div>
    <div
      class="mm-box"
    >
      <h6
        class="mm-box mm-text mm-text--body-sm mm-box--color-text-default"
        data-testid="all-networks-added"
      >
        <span>
           
          You've added all the popular networks. You can discover more networks 
          <a
            class="add-network__edge-case-box__link"
            href="https://chainlist.wtf/"
            rel="noreferrer"
            target="_blank"
          >
            here
            .
          </a>
           Or you can 
          <button
            class="mm-box mm-text mm-button-base mm-button-base--size-md mm-button-primary mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-inverse mm-box--background-color-primary-default mm-box--rounded-pill"
            type="inline"
          >
            <span
              class="mm-box mm-text mm-text--inherit mm-box--color-primary-inverse"
            >
              <h6
                class="mm-box mm-text mm-text--body-sm mm-box--color-info-default"
              >
                add more networks manually
                .
              </h6>
            </span>
          </button>
          
           
        </span>
      </h6>
    </div>
  </div>
</div>
`;
